File: backend/app/api/v1/staff_notifications.py
"""
Staff Notification & Operational Commitment API

Staff Notification = OC 기반
"약속 존재"가 아니라 "운영 리스크 발생 시점"만 노출

Endpoints:
- GET /staff-notifications - Action Queue (노출 대상 OC)
- GET /staff-notifications/{oc_id} - 단일 OC 조회
- POST /staff-notifications/{oc_id}/done - 완료 처리
- POST /staff-notifications/{oc_id}/confirm-resolve - suggested_resolve 확정
- POST /staff-notifications/{oc_id}/reject-resolve - suggested_resolve 거부
- POST /staff-notifications/{oc_id}/confirm-candidate - 후보 확정
- POST /staff-notifications/{oc_id}/reject-candidate - 후보 거부
- GET /conversations/{airbnb_thread_id}/ocs - Conversation의 OC 목록 (Backlog)
"""
from __future__ import annotations

import logging
from datetime import date, datetime
from typing import List, Optional
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel
from sqlalchemy import select
from sqlalchemy.orm import Session

from app.db.session import SessionLocal
from app.domain.models.conversation import Conversation
from app.domain.models.operational_commitment import (
    OperationalCommitment,
    OCStatus,
    OCPriority,
    StaffNotificationItem,
)
from app.services.oc_service import OCService

logger = logging.getLogger(__name__)


# ─────────────────────────────────────────────────────────────
# DB Session
# ─────────────────────────────────────────────────────────────

def get_db():
    db = SessionLocal()
    try:
        yield db
    finally:
        db.close()


# ─────────────────────────────────────────────────────────────
# DTOs
# ─────────────────────────────────────────────────────────────

class OCDTO(BaseModel):
    id: str
    conversation_id: str
    topic: str
    description: str
    evidence_quote: str
    target_time_type: str
    target_date: Optional[str]
    status: str
    resolution_reason: Optional[str]
    resolution_evidence: Optional[str]  # 해소 제안 근거 (게스트 메시지)
    is_candidate_only: bool
    created_at: str
    
    class Config:
        from_attributes = True


class OCListResponse(BaseModel):
    airbnb_thread_id: str
    items: List[OCDTO]
    total: int


class StaffNotificationDTO(BaseModel):
    """Staff Notification 항목 (Action Queue용)"""
    oc_id: str
    conversation_id: str
    airbnb_thread_id: str
    topic: str
    description: str
    evidence_quote: str
    priority: str  # immediate, upcoming, pending
    guest_name: Optional[str]
    checkin_date: Optional[str]
    checkout_date: Optional[str]
    status: str
    resolution_reason: Optional[str]
    resolution_evidence: Optional[str]  # 해소 제안 근거 (게스트 메시지)
    is_candidate_only: bool
    target_time_type: str
    target_date: Optional[str]
    created_at: Optional[str]


class StaffNotificationListResponse(BaseModel):
    items: List[StaffNotificationDTO]
    total: int
    as_of: str  # 조회 기준 날짜


class OCActionResponse(BaseModel):
    oc_id: str
    status: str
    action: str
    success: bool


# ─────────────────────────────────────────────────────────────
# Router
# ─────────────────────────────────────────────────────────────

router = APIRouter(prefix="/staff-notifications", tags=["staff-notifications"])


# ─────────────────────────────────────────────────────────────
# Staff Notification Endpoints
# ─────────────────────────────────────────────────────────────

@router.get("", response_model=StaffNotificationListResponse)
def get_staff_notifications(
    limit: int = Query(50, ge=1, le=200),
    db: Session = Depends(get_db),
):
    """
    Staff Notification Action Queue
    
    지금 처리해야 할 OC 목록 (노출 조건 만족하는 것만)
    
    노출 규칙:
    - status = pending or suggested_resolve
    - explicit: D-1부터 노출
    - implicit: 즉시 노출
    - 체류 중: 최상단 고정
    """
    today = date.today()
    service = OCService(db)
    
    items = service.get_staff_notifications(today=today, limit=limit)
    
    return StaffNotificationListResponse(
        items=[
            StaffNotificationDTO(
                oc_id=str(item.oc_id),
                conversation_id=str(item.conversation_id),
                airbnb_thread_id=item.airbnb_thread_id,
                topic=item.topic,
                description=item.description,
                evidence_quote=item.evidence_quote,
                priority=item.priority.value,
                guest_name=item.guest_name,
                checkin_date=item.checkin_date.isoformat() if item.checkin_date else None,
                checkout_date=item.checkout_date.isoformat() if item.checkout_date else None,
                status=item.status,
                resolution_reason=item.resolution_reason,
                resolution_evidence=item.resolution_evidence,
                is_candidate_only=item.is_candidate_only,
                target_time_type=item.target_time_type,
                target_date=item.target_date.isoformat() if item.target_date else None,
                created_at=item.created_at.isoformat() if item.created_at else None,
            )
            for item in items
        ],
        total=len(items),
        as_of=today.isoformat(),
    )


@router.get("/{oc_id}", response_model=OCDTO)
def get_staff_notification(
    oc_id: UUID,
    db: Session = Depends(get_db),
):
    """단일 OC 조회"""
    from app.repositories.oc_repository import OCRepository
    
    repo = OCRepository(db)
    oc = repo.get_by_id(oc_id)
    
    if not oc:
        raise HTTPException(status_code=404, detail="OC not found")
    
    return OCDTO(
        id=str(oc.id),
        conversation_id=str(oc.conversation_id),
        topic=oc.topic,
        description=oc.description,
        evidence_quote=oc.evidence_quote,
        target_time_type=oc.target_time_type,
        target_date=oc.target_date.isoformat() if oc.target_date else None,
        status=oc.status,
        resolution_reason=oc.resolution_reason,
        is_candidate_only=oc.is_candidate_only,
        created_at=oc.created_at.isoformat() if oc.created_at else None,
    )


# ─────────────────────────────────────────────────────────────
# OC Action Endpoints
# ─────────────────────────────────────────────────────────────

@router.post("/{oc_id}/done", response_model=OCActionResponse)
def mark_oc_done(
    oc_id: UUID,
    db: Session = Depends(get_db),
):
    """
    OC 완료 처리
    
    운영 액션을 수행했을 때 호출
    """
    service = OCService(db)
    oc = service.mark_done(oc_id, by="host")
    
    if not oc:
        raise HTTPException(status_code=404, detail="OC not found")
    
    return OCActionResponse(
        oc_id=str(oc_id),
        status=oc.status,
        action="done",
        success=True,
    )


@router.post("/{oc_id}/confirm-resolve", response_model=OCActionResponse)
def confirm_oc_resolve(
    oc_id: UUID,
    db: Session = Depends(get_db),
):
    """
    suggested_resolve 확정
    
    시스템이 제안한 해소를 운영자가 1클릭으로 확정
    """
    service = OCService(db)
    oc = service.confirm_resolve(oc_id, by="host")
    
    if not oc:
        raise HTTPException(status_code=404, detail="OC not found or not in suggested_resolve status")
    
    return OCActionResponse(
        oc_id=str(oc_id),
        status=oc.status,
        action="confirm_resolve",
        success=True,
    )


@router.post("/{oc_id}/reject-resolve", response_model=OCActionResponse)
def reject_oc_resolve(
    oc_id: UUID,
    db: Session = Depends(get_db),
):
    """
    suggested_resolve 거부
    
    시스템 제안을 거부하고 pending으로 복귀
    """
    service = OCService(db)
    oc = service.reject_resolve(oc_id)
    
    if not oc:
        raise HTTPException(status_code=404, detail="OC not found")
    
    return OCActionResponse(
        oc_id=str(oc_id),
        status=oc.status,
        action="reject_resolve",
        success=True,
    )


@router.post("/{oc_id}/confirm-candidate", response_model=OCActionResponse)
def confirm_oc_candidate(
    oc_id: UUID,
    db: Session = Depends(get_db),
):
    """
    후보 확정
    
    refund/payment/compensation 등 운영자 확정 필요한 OC 확정
    """
    service = OCService(db)
    oc = service.confirm_candidate(oc_id)
    
    if not oc:
        raise HTTPException(status_code=404, detail="OC not found or not a candidate")
    
    return OCActionResponse(
        oc_id=str(oc_id),
        status=oc.status,
        action="confirm_candidate",
        success=True,
    )


@router.post("/{oc_id}/reject-candidate", response_model=OCActionResponse)
def reject_oc_candidate(
    oc_id: UUID,
    db: Session = Depends(get_db),
):
    """
    후보 거부
    
    잘못 추출된 OC 후보 거부
    """
    service = OCService(db)
    oc = service.reject_candidate(oc_id)
    
    if not oc:
        raise HTTPException(status_code=404, detail="OC not found or not a candidate")
    
    return OCActionResponse(
        oc_id=str(oc_id),
        status=oc.status,
        action="reject_candidate",
        success=True,
    )


# ─────────────────────────────────────────────────────────────
# Conversation OC Endpoints (Backlog)
# ─────────────────────────────────────────────────────────────

@router.get("/conversations/{airbnb_thread_id}/ocs", response_model=OCListResponse)
def get_conversation_ocs(
    airbnb_thread_id: str,
    include_resolved: bool = Query(False),
    db: Session = Depends(get_db),
):
    """
    Conversation의 OC 목록 조회
    
    Backlog 화면용 - 전체 OC 조회 가능
    """
    # Conversation 조회
    conv = db.execute(
        select(Conversation).where(Conversation.airbnb_thread_id == airbnb_thread_id)
    ).scalar_one_or_none()
    
    if not conv:
        return OCListResponse(
            airbnb_thread_id=airbnb_thread_id,
            items=[],
            total=0,
        )
    
    service = OCService(db)
    ocs = service.get_conversation_ocs(
        conversation_id=conv.id,
        include_resolved=include_resolved,
    )
    
    return OCListResponse(
        airbnb_thread_id=airbnb_thread_id,
        items=[
            OCDTO(
                id=str(oc.id),
                conversation_id=str(oc.conversation_id),
                topic=oc.topic,
                description=oc.description,
                evidence_quote=oc.evidence_quote,
                target_time_type=oc.target_time_type,
                target_date=oc.target_date.isoformat() if oc.target_date else None,
                status=oc.status,
                resolution_reason=oc.resolution_reason,
                is_candidate_only=oc.is_candidate_only,
                created_at=oc.created_at.isoformat() if oc.created_at else None,
            )
            for oc in ocs
        ],
        total=len(ocs),
    )

File: frontend/src/api/staffNotifications.ts
// src/api/staffNotifications.ts
/**
 * Staff Notification API (OC 기반)
 * 
 * Staff Notification = Operational Commitment 기반
 * "약속 존재"가 아니라 "운영 리스크 발생 시점"만 노출
 */
import { apiGet, apiPost } from "./client";

// ============================================================
// Types
// ============================================================

export type OCPriority = "immediate" | "upcoming" | "pending";
export type OCStatus = "pending" | "done" | "resolved" | "suggested_resolve";

export interface StaffNotificationDTO {
  oc_id: string;
  conversation_id: string;
  airbnb_thread_id: string;  // 백엔드 필드명에 맞춤
  topic: string;
  description: string;
  evidence_quote: string;
  priority: OCPriority;
  guest_name: string | null;
  property_code: string | null;
  property_name: string | null;
  checkin_date: string | null;
  checkout_date: string | null;
  status: OCStatus;
  resolution_reason: string | null;
  resolution_evidence: string | null;  // 해소 제안 근거 (게스트 메시지)
  is_candidate_only: boolean;
  target_time_type: string;
  target_date: string | null;
  created_at: string | null;
}

export interface StaffNotificationListResponse {
  items: StaffNotificationDTO[];
  total: number;
  as_of: string;
}

export interface OCDTO {
  id: string;
  conversation_id: string;
  topic: string;
  description: string;
  evidence_quote: string;
  target_time_type: string;
  target_date: string | null;
  status: OCStatus;
  resolution_reason: string | null;
  resolution_evidence: string | null;  // 해소 제안 근거
  is_candidate_only: boolean;
  created_at: string;
}

export interface OCListResponse {
  airbnb_thread_id: string;  // 백엔드 필드명에 맞춤
  items: OCDTO[];
  total: number;
}

export interface OCActionResponse {
  oc_id: string;
  status: string;
  action: string;
  success: boolean;
}

// ============================================================
// Topic/Priority Labels
// ============================================================

export const OC_TOPIC_LABELS: Record<string, string> = {
  early_checkin: "얼리체크인",
  follow_up: "후속 안내",
  facility_issue: "시설 문제",
  refund_check: "환불 확인",
  payment: "결제",
  compensation: "보상",
};

export const OC_PRIORITY_STYLES: Record<OCPriority, { bg: string; text: string; border: string; label: string }> = {
  immediate: {
    bg: "bg-rose-500/20",
    text: "text-rose-300",
    border: "border-rose-500/40",
    label: "🔴 즉시",
  },
  upcoming: {
    bg: "bg-amber-500/20",
    text: "text-amber-300",
    border: "border-amber-500/40",
    label: "🟡 예정",
  },
  pending: {
    bg: "bg-slate-500/20",
    text: "text-slate-300",
    border: "border-slate-500/40",
    label: "⚪ 대기",
  },
};

export const OC_STATUS_LABELS: Record<OCStatus, string> = {
  pending: "대기",
  done: "완료",
  resolved: "해소",
  suggested_resolve: "해소 제안",
};

// ============================================================
// API Functions
// ============================================================

/**
 * Staff Notification Action Queue 조회
 * GET /api/v1/staff-notifications
 */
export async function fetchStaffNotifications(params?: {
  limit?: number;
}): Promise<StaffNotificationListResponse> {
  return apiGet<StaffNotificationListResponse>("/staff-notifications", params);
}

/**
 * 단일 OC 조회
 * GET /api/v1/staff-notifications/{oc_id}
 */
export async function getOC(ocId: string): Promise<OCDTO> {
  return apiGet<OCDTO>(`/staff-notifications/${ocId}`);
}

/**
 * Conversation의 OC 목록 조회 (Backlog)
 * GET /api/v1/staff-notifications/conversations/{thread_id}/ocs
 */
export async function getConversationOCs(
  threadId: string,
  params?: { include_resolved?: boolean }
): Promise<OCListResponse> {
  return apiGet<OCListResponse>(
    `/staff-notifications/conversations/${threadId}/ocs`,
    params
  );
}

/**
 * OC 완료 처리
 * POST /api/v1/staff-notifications/{oc_id}/done
 */
export async function markOCDone(ocId: string): Promise<OCActionResponse> {
  return apiPost<OCActionResponse>(`/staff-notifications/${ocId}/done`, {});
}

/**
 * suggested_resolve 확정
 * POST /api/v1/staff-notifications/{oc_id}/confirm-resolve
 */
export async function confirmOCResolve(ocId: string): Promise<OCActionResponse> {
  return apiPost<OCActionResponse>(`/staff-notifications/${ocId}/confirm-resolve`, {});
}

/**
 * suggested_resolve 거부
 * POST /api/v1/staff-notifications/{oc_id}/reject-resolve
 */
export async function rejectOCResolve(ocId: string): Promise<OCActionResponse> {
  return apiPost<OCActionResponse>(`/staff-notifications/${ocId}/reject-resolve`, {});
}

/**
 * 후보 확정
 * POST /api/v1/staff-notifications/{oc_id}/confirm-candidate
 */
export async function confirmOCCandidate(ocId: string): Promise<OCActionResponse> {
  return apiPost<OCActionResponse>(`/staff-notifications/${ocId}/confirm-candidate`, {});
}

/**
 * 후보 거부
 * POST /api/v1/staff-notifications/{oc_id}/reject-candidate
 */
export async function rejectOCCandidate(ocId: string): Promise<OCActionResponse> {
  return apiPost<OCActionResponse>(`/staff-notifications/${ocId}/reject-candidate`, {});
}

/**
 * 우선순위 변경
 * POST /api/v1/staff-notifications/{oc_id}/change-priority
 * 
 * - immediate: target_date = today
 * - upcoming: target_date = today + 1
 * - pending: target_date 유지
 */
export async function changeOCPriority(
  ocId: string, 
  priority: OCPriority
): Promise<OCActionResponse> {
  return apiPost<OCActionResponse>(`/staff-notifications/${ocId}/change-priority`, { priority });
}
